core/commonMain/src/kotlinx/serialization/Serializers.kt

```diff
@@ -156,14 +165,35 @@ private fun SerializersModule.serializerByKTypeImpl(
     } else {
         val serializers = serializersForParameters(typeArguments, failOnMissingTypeArgSerializer) ?: return null
         // first, we look among the built-in serializers, because the parameter could be contextual
-        rootClass.parametrizedSerializerOrNull(typeArguments, serializers) ?: getContextual(
-            rootClass,
-            serializers
-        )
+        rootClass.parametrizedSerializerOrNull(serializers) { typeArguments[0].classifier }
+            ?: getContextual(
+                rootClass,
+                serializers
+            )
     }
     return contextualSerializer?.cast<Any>()?.nullable(isNullable)
 }
 
+@OptIn(ExperimentalSerializationApi::class)
+private fun SerializersModule.serializerByKClassImpl(
+    rootClass: KClass<Any>,
+    typeSerializers: List<KSerializer<Any?>>,
+    isNullable: Boolean
+): KSerializer<Any?>? {
+
+    val serializer = if (typeSerializers.isEmpty()) {
+        rootClass.serializerOrNull() ?: getContextual(rootClass)
+    } else {
+        rootClass.parametrizedSerializerOrNull(typeSerializers) { throw SerializationException("Array serializer can not be create from KClass arguments") }
```

Comment:
This error message is cryptic: it doesn't explain what is 'KClass arguments', as we have only one KClass<Any>; nor it does explain how to fix it